rpcs3/Emu/RSX/Common/BufferUtils.cpp

```diff
@@ -755,36 +756,286 @@ namespace
 
 	struct primitive_restart_impl
 	{
+#if __AVX2__
+		static
+		std::tuple<u16, u16, u32> upload_u16_swapped_avx2(const void *src, void *dst, u32 count, u32 restart_index)
+		{
+			u32 dst_index = 0;
+
+			const __m256i mask = _mm256_set_epi8(
+				0x1E, 0x1F, 0x1C, 0x1D,
+				0x1A, 0x1B, 0x18, 0x19,
+				0x16, 0x17, 0x14, 0x15,
+				0x12, 0x13, 0x10, 0x11,
+				0xE, 0xF, 0xC, 0xD,
+				0xA, 0xB, 0x8, 0x9,
+				0x6, 0x7, 0x4, 0x5,
+				0x2, 0x3, 0x0, 0x1);
+
+			auto src_stream = (const __m256i*)src;
+			auto dst_stream = (__m256i*)dst;
+
+			__m256i restart = _mm256_set1_epi16(restart_index);
+			__m256i min = _mm256_set1_epi16(0xffff);
+			__m256i max = _mm256_set1_epi16(0);
+
+			const auto iterations = count / 8;
+			for (unsigned n = 0; n < iterations; ++n)
+			{
+				const __m256i raw = _mm256_loadu_si256(src_stream++);
+				const __m256i value = _mm256_shuffle_epi8(raw, mask);
+				const __m256i mask = _mm256_cmpeq_epi16(restart, value);
+				const __m256i tmp = _mm256_andnot_si256(mask, value);
+				max = _mm256_max_epu16(max, tmp);
+				min = _mm256_min_epu16(min, value);
+				_mm256_storeu_si256(dst_stream++, value);
+			}
+
+			const __m256i mask_step1 = _mm256_set_epi8(
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0x1F, 0x1E, 0x1D, 0x1C, 0x1B, 0x1A, 0x19, 0x18,
+				0x17, 0x16, 0x15, 0x14, 0x13, 0x12, 0x11, 0x10);
+
+			const __m256i mask_step2 = _mm256_set_epi8(
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0xF, 0xE, 0xD, 0xC, 0xB, 0xA, 0x9, 0x8);
+
+			const __m256i mask_step3 = _mm256_set_epi8(
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0x7, 0x6, 0x5, 0x4);
+
+			const __m256i mask_step4 = _mm256_set_epi8(
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0x3, 0x2);
+
+			__m256i tmp = _mm256_shuffle_epi8(min, mask_step1);
+			min = _mm256_min_epu16(min, tmp);
+			tmp = _mm256_shuffle_epi8(min, mask_step2);
+			min = _mm256_min_epu16(min, tmp);
+			tmp = _mm256_shuffle_epi8(min, mask_step3);
+			min = _mm256_min_epu16(min, tmp);
+			tmp = _mm256_shuffle_epi8(min, mask_step4);
+			min = _mm256_min_epu16(min, tmp);
+
+			tmp = _mm256_shuffle_epi8(max, mask_step1);
+			max = _mm256_max_epu16(max, tmp);
+			tmp = _mm256_shuffle_epi8(max, mask_step2);
+			max = _mm256_max_epu16(max, tmp);
+			tmp = _mm256_shuffle_epi8(max, mask_step3);
+			max = _mm256_max_epu16(max, tmp);
+			tmp = _mm256_shuffle_epi8(max, mask_step4);
+			max = _mm256_max_epu16(max, tmp);
+
+			const u16 min_index = u16(_mm_cvtsi128_si32(_mm256_castsi256_si128(min)) & 0xFFFF);
+			const u16 max_index = u16(_mm_cvtsi128_si32(_mm256_castsi256_si128(max)) & 0xFFFF);
```

Comment:
Can simply use _mm256_extract_epi16.

Reply:
Do you know if the latency and throughput are equal or better for this one?

Reply:
My version generates a `vmovd`, while your version generates a `vpextrw`.

I haven’t been able to find information on latency and throughput about them, but I believe a simple `mov` will always be more efficient.

Reply:
well, it wasn't because of performance but more because of readablity.

Reply:
The latest version of the AVX2 commit does the conversion from `ymm` to `xmm` earlier, so there isn’t any readability issue anymore, thanks for the suggestion anyway. :)